src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2943,6 +2945,12 @@ module Crystal
     end
 
     def visit(node : StringInterpolation)
+      # Eagerly solve Paths before expansion to see if
+      # we can inline some content.
+      node.expressions.each do |exp|
+        exp.accept self if exp.is_a?(Path)
+      end
+
```

Comment:
This is what I worry about because it can change the evaluation order of certain expressions:

```crystal
class Foo
end

class Bar(T) < Foo
  def initialize(x : T)
  end
end

def foo
  Bar.new(1)
  nil
end

X = {{ Foo.subclasses.join("") }}

# "Bar(T)Bar(Int32)" before this change
"#{foo}#{X}" # => "Bar(T)"
```

Some other nodes like `Case` also evaluate subnodes out of order. It is up to you to decide whether this counts as a breaking change.

Reply:
Good point. If that's the case, I'd like to remove this kind of reflection from Crystal as it's very unreliable.

Reply:
Another idea would be to check this optimization right before codegen. I'll try that later.

Reply:
@HertzDevil Done! Now your example still gives the same result as before.

I actually like this solution better because it doesn't modify the main semantic phase.